### DISCUSSION TITLE
Falcon Performance discussion

### DISCUSSION THREAD

**Discussion by gr790** [socketry/falcon#141] [General]
Hello,

I am evaluating Falcon as an solution to support HTTP2 web server in Production. As of now I am thinking (Falcon + Rails), to support HTTP2 server and processing logic is implemented in rails framework, it nothing fancy stuff, just some conversion module from 1 format of http request to another format of requests and vice versa. 

I am able to run it as you instructed in README on github.com. Even I see my rails routes/controller and library methods are getting called if I am sending traffic on configured port and IP. My library methods just extract the headers and return 200 OK.

I tried to run the performance using hey on local server, but it seems there are some bottle necks which leads to lower performance. 

This is how I am running falcon.
`falcon serve --threaded --count 40 --port 8081` ... Do you see any improvements in number of threads?

This is how I am sending traffic.
`./bin/hey_linux_amd64 -n 10000 -c 100 -h2 -m POST -D ../sample.json https://localhost:8081/services/sample`. I am expecting there would be 100 clients now...

This is the number I am getting.

```
Summary:
  Total:        111.7138 secs
  Slowest:      2.6681 secs
  Fastest:      0.0115 secs
  Average:      1.1003 secs
  Requests/sec: 89.5145


Response time histogram:
  0.011 [1]     |
  0.277 [346]   |■■■
  0.543 [951]   |■■■■■■■■■
  0.808 [363]   |■■■
  1.074 [1855]  |■■■■■■■■■■■■■■■■■
  1.340 [4355]  |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  1.605 [1631]  |■■■■■■■■■■■■■■■
  1.871 [376]   |■■■
  2.137 [87]    |■
  2.402 [30]    |
  2.668 [5]     |


Latency distribution:
  10% in 0.4085 secs
  25% in 1.0177 secs
  50% in 1.1447 secs
  75% in 1.3070 secs
  90% in 1.4743 secs
  95% in 1.6045 secs
  99% in 1.9016 secs

Details (average, fastest, slowest):
  DNS+dialup:   0.0000 secs, 0.0115 secs, 2.6681 secs
  DNS-lookup:   0.0002 secs, 0.0000 secs, 0.0611 secs
  req write:    0.4683 secs, 0.0000 secs, 1.7996 secs
  resp wait:    0.5896 secs, 0.0112 secs, 1.4613 secs
  resp read:    0.0391 secs, 0.0000 secs, 1.0381 secs

Status code distribution:
  [204] 10000 responses
```

- Do you suggest any tools to check the performance ?
- Do you see any improvements tweaks to supports atleast 150+ TPS ?, I don't want to stress the CPU above 60%. 

Looking +ve/quick response.
Thanks.

**Comment by gr790** [comment]
Hello,

I tried to use forked mode.
`falcon serve --forked --count 2 --port 8081`

And running another benchmark tool 
`./wrk -t4 -c50 -d300s https://localhost:8081/services/sample`

This is the number I am getting
`
  65006 requests in 5.00m, 32.43MB read
  Non-2xx or 3xx responses: 65006
Requests/sec:    216.63
Transfer/sec:    110.66KB
`

Thanks.

**Comment by ioquatix** [comment]
Adding more threads than processors will probably make it slower.

Using threaded mode on MRI is slower than forked.

Do the benchmark tools you use support HTTP/2?

Obviously it will depend on hardware, but maybe compare it with puma to see if there are performance regressions in falcon.

Also, what version of rails are you using? It's not the fastest framework.
